### PR TITLE
Extend rules docs

### PR DESCRIPTION
**Pull request by malteneuss**
Part of #904

### CODE REVIEW

File: lib/Hakyll/Core/Rules.hs [hunk 1]
```diff
@@ -1,19 +1,32 @@
...
+1. Inputs (like Markdown files) to process (collected with e.g. 'match' or 'create').
+2. Transformation steps (like Markdown to HTML) to compute ("compile") for each input its output content
+(collected with 'compile' and 'Hakyll.Core.Compiler.Compiler').
+3. Routing to determine to which file an output content will be written out. For a static site
+this translates to determine under which URL the output content will be accessible
+(configured with 'route' and 'Hakyll.Core.Routes.Routes').
+
+A typical usage example looks as follows:
+
+> {-# LANGUAGE OverloadedStrings #-}    -- write 'match "posts/**.md"' instead of 'match $ fromGlob "posts/**.md"'
```
Comment (malteneuss) [inline lib/Hakyll/Core/Rules.hs:15]: I can't get this extension to show up in the docs. Any ideas?

File: lib/Hakyll/Core/Rules.hs [hunk 2]
```diff
@@ -130,18 +143,100 @@ matchInternal pattern getIDs rules = do
     setMatches ids env = env {rulesMatches = ids}
 
 --------------------------------------------------------------------------------
-match :: Pattern -> Rules () -> Rules ()
+{- | Add a selection of which input files to process (using the given
+[glob pattern](https://w.wiki/4Wsx)) to the given remaining 'Rules' pipeline.
```
Comment (malteneuss) [inline lib/Hakyll/Core/Rules.hs:147]: I can't get this link to render correctly in the docs. Any ideas?

File: lib/Hakyll/Core/Rules.hs [hunk 3]
```diff
@@ -130,18 +143,100 @@ matchInternal pattern getIDs rules = do
...
+match :: Pattern  -- ^ Glob pattern
+      -> Rules () -- ^ Remaining processing pipeline
+      -> Rules () -- ^ Resulting pipeline
 match pattern = matchInternal pattern $ getMatches pattern
 
 
 --------------------------------------------------------------------------------
-matchMetadata :: Pattern -> (Metadata -> Bool) -> Rules () -> Rules ()
+{- | Add a selection of which input files to process (using the given
+[glob pattern](https://w.wiki/4Wsx) and metadata predicate) to the given remaining 'Rules' pipeline.
```
Comment (malteneuss) [inline lib/Hakyll/Core/Rules.hs:178]: same here

File: lib/Hakyll/Core/Rules.hs (hunk 1 above)
Comment (Minoru) [reply lib/Hakyll/Core/Rules.hs:15]: Apparently one can't nest block comments inside Haddock block comments. You'll have to switch to outer one:

```
-- | Blah blah
--
-- > {-# LANGUAGE OverloadedStrings #-}
```

File: lib/Hakyll/Core/Rules.hs (hunk 2 above)
Comment (Minoru) [reply lib/Hakyll/Core/Rules.hs:147]: Haddock has its own markup for links: https://www.haskell.org/haddock/doc/html/ch03s08.html#idm140354810727840

Also, are there any reason to use a URL shortener here? Why not link to Wikipedia directly?

File: lib/Hakyll/Core/Rules.hs (hunk 1 above)
Comment (malteneuss) [reply lib/Hakyll/Core/Rules.hs:15]: Thanks. I added the workaround. Here's the corresponding bug ticket for Haddock https://github.com/haskell/haddock/issues/1254

File: lib/Hakyll/Core/Rules.hs (hunk 2 above)
Comment (malteneuss) [reply lib/Hakyll/Core/Rules.hs:147]: Markdown style links are supported https://haskell-haddock.readthedocs.io/en/latest/markup.html#links (the docs on haskell.org seem to be terribly outdated). I got it working.

I used a shortener to test if the parenthesis in the url `https://en.wikipedia.org/wiki/Glob_(programming)` are a problem; they are, but can be escaped. 

Apparently, there's another bug/quirk in Haddock:
> {- | Add a selection...
> given [glob pattern](https://en.wikipedia.org/wiki/Glob_(programming\))) ...

renders correctly, but having such a link at the beginning of a block comment line doesn't:
> {- | Add a selection ...
> [glob pattern](https://en.wikipedia.org/wiki/Glob_(programming\))) to the given remaining 'Rules' pipeline.

I'll add a bug report on Haddock.

File: lib/Hakyll/Core/Rules.hs (hunk 3 above)
Comment (malteneuss) [reply lib/Hakyll/Core/Rules.hs:178]: Resolved.

**Comment by malteneuss** [comment]
Any review comments? I would like to finish this PR soon.

**Comment by Minoru** [comment]
Oof, sorry, I didn't realize that I've been postponing this for 4 months now! I'll make time for a review tomorrow.

In the meantime, it'd be nice if you could rebase this onto the latest `master` to fix CI here.

File: lib/Hakyll/Core/Rules.hs [hunk 4]
```diff
@@ -160,19 +335,117 @@ version v rules = do
...
+>
+> -- To Hakyll templates are just plain files that have to be processed
+> -- and placed into the store like any other file (but without routing).
+> -- e.g. file on disk: 'templates/post.html'
+> match "templates/*" $ compile templateBodyCompiler
+Note how a Markdown post that is compiled to HTML using
+'Hakyll.Web.Pandoc.pandocCompiler' in a first step and then embedded into
+a HTMl 'Hakyll.Web.Template.Template' in a second step by using
+'Hakyll.Web.Template.loadAndApplyTemplate'.
+We can use templates to to control the design and layout of a webpage.
```
Comment (Minoru) [inline lib/Hakyll/Core/Rules.hs:388]:
```suggestion
We can use templates to control the design and layout of a webpage.
```

File: lib/Hakyll/Core/Rules.hs [hunk 5]
```diff
@@ -160,19 +335,117 @@ version v rules = do
...
 
 --------------------------------------------------------------------------------
--- | Add a route.
---
--- This adds a route for all items matching the current pattern.
-route :: Routes -> Rules ()
+{- | Add (or replace) routing in the current 'Rules' value.
+__This functions controls if and where the compiled results are written out__
+(use one of the 'match' functions to control what content is processed and
+'compile' to control how).
```
Comment (Minoru) [inline lib/Hakyll/Core/Rules.hs:405]: For consistency with other parts of the docs:

```suggestion
__This functions controls IF and WHERE the compiled results are written out__
(use one of the 'match' functions to control WHAT content is processed and
'compile' to control HOW).
```

File: lib/Hakyll/Core/Rules.hs [hunk 6]
```diff
@@ -160,19 +335,117 @@ version v rules = do
...
+See 'Hakyll.Core.Routes.Routes' and 'Hakyll.Core.Identifier.Identifier' for
+details on how output filepaths are computed.
+
+Hint:
+__If there's no route attached to a rule, the compilation result is not written out__.
+However, the compilation result saved to the 'Hakyll.Core.Store.Store'
```
Comment (Minoru) [inline lib/Hakyll/Core/Rules.hs:411]:
```suggestion
However, the compilation result is saved to the 'Hakyll.Core.Store.Store'
```

File: lib/Hakyll/Core/Rules.hs [hunk 7]
```diff
@@ -160,19 +335,117 @@ version v rules = do
...
+> -- Rule 2 (with routing)
+> match "posts/**.md" $ do
+>     route $ setExtension "html"
+>     compile $ do
+>        -- load compiled result of other rule with explicit identifier.
+>        postTemplate <- loadBody "templates/post.html"
+>        pandocCompiler >>= applyTemplate postTemplate defaultContext
+Note that we don't set a route in the first rule to avoid writing out our
+compiled templates (a webpage containing raw
+'Hakyll.Web.Template.templateCompiler' templates makes rarely sense after all).
```
Comment (Minoru) [inline lib/Hakyll/Core/Rules.hs:434]: Not sure what you meant here. Perhaps rewrite it this way?

```suggestion
'Hakyll.Web.Template.templateCompiler' templates rarely makes sense, after all).
```

File: lib/Hakyll/Core/Rules.hs [hunk 8]
```diff
@@ -1,19 +1,43 @@
...
--- A typical usage example would be:
---
+-- | This module provides a declarative Domain Specific Language (DSL) to
+-- generate a static site by specifying transformation 'Rules' (although the
+-- use case is not limited to static sites).
+-- Each rule normally consists of three parts:
+-- 
+-- 1. Source files (like Markdown files) to process (collected with e.g.
+-- 'match' or 'create').
+-- 2. Compilation steps (like Markdown to HTML) to transform a files' content
```
Comment (Minoru) [inline lib/Hakyll/Core/Rules.hs:9]:
```suggestion
-- 2. Compilation steps (like Markdown to HTML) to transform files' content
```

("Files" is plural, so we can't use singular "a" here.)

**Review by Minoru — APPROVED** [review]
Thanks for your patience! I think this looks good. I fixed a couple typos, and in one place I'm not sure what you meant — please commit my suggestions if you agree with them, or let's discuss how to rewrite those parts better.

File: lib/Hakyll/Core/Rules.hs (hunk 7 above)
Comment (malteneuss) [reply lib/Hakyll/Core/Rules.hs:434]: I wanted to explain that it doesn't makes sense to render raw templates to an output file; that's why there's no routing part in the first rule. Maybe just leave the text inside parenthesis out completely?

Comment (malteneuss) [reply lib/Hakyll/Core/Rules.hs]: yeah, let's just remove it.

**Comment by Minoru** [comment]
Cool! Thank you very much for the work and especially for the patience ^_^ Merging!

**Pull request merged**